### PR TITLE
JBIDE-20366: Check for missing m2e connectors after import

### PR DESCRIPTION
**Pull request by mickaelistria**

### CODE REVIEW

**Comment by alexeykazakov** [comment]
needs @fbricon rivew

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java [hunk 1]
```diff
@@ -32,16 +34,61 @@
...
+import org.eclipse.m2e.core.ui.internal.wizards.LifecycleMappingDiscoveryHelper;
+import org.eclipse.m2e.core.ui.internal.wizards.MappingDiscoveryJob;
+import org.eclipse.ui.internal.wizards.datatransfer.EasymportJob;
 import org.eclipse.ui.wizards.datatransfer.ProjectConfigurator;
 
 public class MavenProjectConfigurator implements ProjectConfigurator {
 	
+	private static class CumulativeMappingDiscoveryJob extends MappingDiscoveryJob {
+		private static CumulativeMappingDiscoveryJob INSTANCE;
+		private Set<IProject> toProcess;
```
Comment (fbricon) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: memory leak: toProcess keeps project references in job instance, after import, and keeps on piling new projects

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java [hunk 2]
```diff
@@ -32,16 +34,61 @@
...
 import org.eclipse.m2e.core.project.ResolverConfiguration;
+import org.eclipse.m2e.core.ui.internal.wizards.LifecycleMappingDiscoveryHelper;
+import org.eclipse.m2e.core.ui.internal.wizards.MappingDiscoveryJob;
+import org.eclipse.ui.internal.wizards.datatransfer.EasymportJob;
 import org.eclipse.ui.wizards.datatransfer.ProjectConfigurator;
 
 public class MavenProjectConfigurator implements ProjectConfigurator {
 	
+	private static class CumulativeMappingDiscoveryJob extends MappingDiscoveryJob {
+		private static CumulativeMappingDiscoveryJob INSTANCE;
```
Comment (fbricon) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: does it really need to be a singleton? It's not heavy to instanciate and the job will be kept in memory forever after the 1st run.

Comment (mickaelistria) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: The constraints are:
* MavenProjectConfigurator is stateless (its lifecycle and the multiplicity aren't specified, and I'd rather avoid to constrain them). 
* UpdateMavenConfigurationJob has multiple instances whom orchestration isn't something we can control
* While importing, we don't want the Mapping discovery dialog to pop-up before the import isn't completed, so we have to make sure only one instance will be used for an import operation. 

Basically, we need a single stateful operation, after multiple stateless ones. The singleton job and a join on the family is the best thing I found to allow that.

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java (hunk 1 above)
Comment (mickaelistria) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: Thanks for noticing it. I'll fix it as part of next iteration.
More than the small memory leak (which is actually size of a pointer * max number of projects in workspace), I believe this can have an important impact on the performance of the connector discovery operations.

**Comment by mickaelistria** [comment]
Updated patch to clear the toProcess set.

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java [hunk 3]
```diff
@@ -32,16 +34,61 @@
...
+				return Status.OK_STATUS;
+			}
+			//Some errors were detected
+			discoverProposals(discoveryRequest, monitor);
+			openProposalWizard(toProcess, discoveryRequest);
+			this.toProcess.clear();
+			return Status.OK_STATUS;
+		}
+			
+		public void addProjects(Collection<IProject> projects) {
```
Comment (fbricon) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: since this is a public method, I'd recommend to check for null input

Comment (mickaelistria) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: OK, done.

**Comment by mickaelistria** [comment]
The latest version of this patch adds a null check.

**Comment by mickaelistria** [comment]
@fbricon : could you please merge it if it's fine for you (both branch and master) ?

**Comment by mickaelistria** [comment]
@fbricon Any objection to merge it in both branch and master (I can do it by myself, but would like someone to +1 )

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java [hunk 4]
```diff
@@ -32,16 +34,63 @@
...
+			if (projects != null) {
```
Comment (fbricon) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: I think this method should check the job status before adding new projects. If the instance is already running, gets past openProposalWizard(), projects will be added to the set that'll be cleared right after.
So basically you should throw an IllegalStateException if the job is running while you're adding projects

**Comment by mickaelistria** [comment]
Updated change as advised.

**Comment by mickaelistria** [comment]
@fbricon can you please review and merge before code freeze?

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java [hunk 5]
```diff
@@ -32,16 +35,72 @@
...
+			this.toProcess.clear();
+			return Status.OK_STATUS;
+		}
+			
+		public void addProjects(Collection<IProject> projects) {
+			synchronized (this.toProcess) {
+				if (this.started) {
+					throw new IllegalStateException("Cannot add projects when processing is started");
+				}
+				if (projects != null) {
```
Comment (fbricon) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: projects can't be null, it's instantiated in the constructor of this singleton

File: plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java [hunk 6]
```diff
@@ -32,16 +35,72 @@
...
+			}
+			//Detect and resolve Lifecycle Mapping issues
+			LifecycleMappingDiscoveryRequest discoveryRequest = LifecycleMappingDiscoveryHelper.createLifecycleMappingDiscoveryRequest(toProcess, monitor);
+			if(discoveryRequest.isMappingComplete()) {
+				return Status.OK_STATUS;
+			}
+			//Some errors were detected
+			discoverProposals(discoveryRequest, monitor);
+			openProposalWizard(toProcess, discoveryRequest);
+			this.toProcess.clear();
```
Comment (fbricon) [inline plugins/org.jboss.tools.playground.easymport.maven/src/org/jboss/tools/playground/easymport/maven/MavenProjectConfigurator.java]: if an exception is thrown, toProcess won't be cleared, leaving you with leaked project references. Should clear in a global finally block

**Comment by fbricon** [comment]
Ok to apply once toProcess.clear() is moved to a try{} finally{} block

**Comment by mickaelistria** [comment]
Done and merged.
